core/editor/src/tools/ellipse.rs

```diff
@@ -1,12 +1,54 @@
-use crate::events::Event;
-use crate::tools::Tool;
+use crate::events::MouseKeys;
+use crate::events::{Event, Response};
+use crate::tools::{Fsm, Tool};
+use crate::Document;
 use document_core::Operation;
 
 #[derive(Default)]
-pub struct Ellipse;
+pub struct Ellipse {
+	state: EllipseToolState,
+}
 
 impl Tool for Ellipse {
-	fn handle_input(&mut self, event: Event) -> Vec<Operation> {
-		todo!();
+	fn handle_input(&mut self, event: &Event, document: &Document) -> (Vec<Response>, Vec<Operation>) {
+		let mut responses = Vec::new();
+		let mut operations = Vec::new();
+		self.state = self.state.transition(event, document, &mut responses, &mut operations);
+
+		(responses, operations)
+	}
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum EllipseToolState {
+	Ready,
+	LmbDown,
+	TransformSelected,
+}
+
+impl Default for EllipseToolState {
+	fn default() -> Self {
+		EllipseToolState::Ready
+	}
+}
+
+impl Fsm for EllipseToolState {
+	fn transition(self, event: &Event, document: &Document, responses: &mut Vec<Response>, operations: &mut Vec<Operation>) -> Self {
+		match (self, event) {
+			(EllipseToolState::Ready, Event::MouseDown(mouse_state)) if mouse_state.mouse_keys.contains(MouseKeys::LEFT) => EllipseToolState::LmbDown,
+
+			// TODO - Check for left mouse button
```

Comment:
don't postpone a simple thing like this, just add the check analogous to the previous match arm

Reply:
That would not work, because the mouse state contains the currently pressed buttons and upon releasing the button, the button in question is no longer pressed. We would have to compute the difference between this and the last mouse state to get the pressed button which would likely happen in the dispatcher. We could also add individual events for LMBDown, LMBUP, etc.

Reply:
Ok, I didn't think about that. I think the latter makes the most sense but that can also be done in a later PR